File: apk/app/src/main/java/com/example/momentra/ui/shell/empty/business/CompanyJoinLink.kt
package com.example.momentra.ui.shell.empty.business

import android.net.Uri

/**
 * Company invite deeplinks — `momentra://c/{code}` / `momentra.app/c/{code}`.
 * Bare codes are only accepted via [parseTyped] (Company Setup text entry).
 */
object CompanyJoinLink {
    private val SHORT = Regex("^[a-hj-np-z2-9]{8}$", RegexOption.IGNORE_CASE)

    fun displayPath(code: String): String =
        "momentra.app/c/${code.trim().lowercase()}"

    fun qrPayload(code: String): String =
        "momentra://c/${code.trim().lowercase()}"

    fun parse(raw: String?): String? {
        if (raw.isNullOrBlank()) return null
        val trimmed = raw.trim()
        extractFromSlashes(trimmed)?.let { return it }

        val uri = runCatching { Uri.parse(trimmed) }.getOrNull() ?: return null
        val scheme = uri.scheme?.lowercase().orEmpty()
        val host = uri.host?.lowercase().orEmpty()
        val segments = uri.pathSegments
        val candidate = when {
            scheme == "momentra" && (host == "c" || host == "company") ->
                segments.lastOrNull()
            (host == "momentra.app" || host == "www.momentra.app") &&
                segments.isNotEmpty() &&
                (segments[0].equals("c", true) || segments[0].equals("company", true)) ->
                segments.getOrNull(1)
            scheme == "momentra" &&
                (segments.getOrNull(0).equals("c", true) || segments.getOrNull(0).equals("company", true)) ->
                segments.getOrNull(1)
            else -> null
        }
        return sanitize(candidate)
    }

    fun parseTyped(raw: String?): String? =
        parse(raw) ?: sanitize(raw?.trim()?.lowercase())

    private fun extractFromSlashes(raw: String): String? {
        val parts = raw.substringBefore('?').substringBefore('#').split('/')
        val marker = parts.indexOfFirst { it.equals("c", true) || it.equals("company", true) }
        if (marker >= 0) return sanitize(parts.getOrNull(marker + 1))
        return null
    }

    private fun sanitize(raw: String?): String? {
        val value = raw?.trim()?.lowercase()?.trim('/') ?: return null
        return value.takeIf { SHORT.matches(it) }
    }
}

sealed class InviteJoinKind {
    data class Group(val code: String) : InviteJoinKind()
    data class Company(val code: String) : InviteJoinKind()
}

object InviteJoinLink {
    fun parse(raw: String?): InviteJoinKind? {
        CompanyJoinLink.parse(raw)?.let { return InviteJoinKind.Company(it) }
        com.example.momentra.ui.shell.empty.group.GroupJoinLink.parse(raw)?.let {
            return InviteJoinKind.Group(it)
        }
        return null
    }
}

File: momentra/momentra/Shell/BusinessEmpty/CompanyJoinLink.swift
import Foundation

/// Company invite deeplinks — `momentra://c/{code}` / `momentra.app/c/{code}`.
/// Does not accept bare codes (those stay GroupJoinLink) to avoid colliding with group invites.
enum CompanyJoinLink {
    private static let shortPattern = #"^[a-hj-np-z2-9]{8}$"#

    static func displayPath(code: String) -> String {
        "momentra.app/c/\(code.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())"
    }

    static func qrPayload(code: String) -> String {
        "momentra://c/\(code.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())"
    }

    static func parse(_ raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        if let fromSlash = extractFromSlashes(trimmed) { return fromSlash }
        guard let url = URL(string: trimmed) else { return nil }
        let scheme = url.scheme?.lowercased() ?? ""
        let host = url.host?.lowercased() ?? ""
        let parts = url.path.split(separator: "/").map(String.init)

        if scheme == "momentra" {
            if host == "c" || host == "company" {
                return sanitize(parts.last)
            }
            if parts.first == "c" || parts.first == "company" {
                return sanitize(parts.dropFirst().first)
            }
        }
        if host == "momentra.app" || host == "www.momentra.app",
           let first = parts.first, first == "c" || first == "company" {
            return sanitize(parts.dropFirst().first)
        }
        return nil
    }

    /// Accept bare 8-char codes when the user typed them into Company Setup.
    static func parseTyped(_ raw: String) -> String? {
        if let linked = parse(raw) { return linked }
        return sanitize(raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())
    }

    private static func extractFromSlashes(_ raw: String) -> String? {
        let parts = raw.split(separator: "?").first.map(String.init) ?? raw
        let segs = parts.split(separator: "#").first.map(String.init) ?? parts
        let pieces = segs.split(separator: "/").map(String.init)
        if let marker = pieces.firstIndex(where: { $0.lowercased() == "c" || $0.lowercased() == "company" }) {
            return sanitize(pieces.dropFirst(marker + 1).first)
        }
        return nil
    }

    private static func sanitize(_ raw: String?) -> String? {
        guard let raw else { return nil }
        let value = raw.trimmingCharacters(in: CharacterSet(charactersIn: "/")).lowercased()
        if value.range(of: shortPattern, options: [.regularExpression, .caseInsensitive]) != nil { return value }
        return nil
    }
}

enum InviteJoinKind {
    case group(String)
    case company(String)
}

enum InviteJoinLink {
    /// Prefer company `/c/` markers, then group `/j/` / bare short codes.
    static func parse(_ raw: String) -> InviteJoinKind? {
        if let company = CompanyJoinLink.parse(raw) { return .company(company) }
        if let group = GroupJoinLink.parse(raw) { return .group(group) }
        return nil
    }
}
